src/eterna/ui/Toolbar.ts

```diff
@@ -782,6 +788,31 @@ export default class Toolbar extends ContainerObject {
             this.addObject(this.submitButton, this.lowerToolbarLayout);
         }
 
+        // kkk add 3d validation button
+        if (this._type === ToolbarType.PUZZLEMAKER) {
+            this.lowerToolbarLayout.addHSpacer(SPACE_WIDE);
+            this.validate3DButton = new ToolbarButton()
+                .up(Bitmaps.ImgFileOpen)
+                .over(Bitmaps.ImgFileOpenHover)
+                .down(Bitmaps.ImgFileOpen)
+                .tooltip('Validate 3D Models');
+
+            this.addObject(this.validate3DButton, this.lowerToolbarLayout);
+            this.regs.add(this.validate3DButton.clicked.connect(() => {
+                fileDialog({accept: ['.cif']}).then((file) => {
+                    const mode: PuzzleEditMode = this.mode as PuzzleEditMode;
+                    const sequence = mode.getSequence().split(' ')[0];
+                    Mol3DGate.checkModelFile(file[0], mode.getSequence().split(' ')[0]).then((resCount:number) => {
+                        if (mode && resCount === sequence.length) mode.add3DSprite(file[0], mode.structure);
+                        else {
+                            const PROMPT = 'Your selected file is mismatched with the puzzle.';
+                            mode?.showDialog(new ErrorDialog(PROMPT));
+                        }
+                    });
+                });
+            }));
```

Comment:
As mentioned earlier, the handler should go in GameMode and be registered in the relevant subclasses.